### PR TITLE
Metric learning train (2)

### PR DESCRIPTION
**Pull request by AlekseySh**
## Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contribution guide](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md)?
- [x] Did you check the code style? `catalyst-make-codestyle && catalyst-check-codestyle` (`pip install -U catalyst-codestyle`).
- [x] Did you make sure to update the docs? We use Google format for all the methods and classes.
- [x] Did you check the docs with `make check-docs`?
- [x] Did you write any new necessary tests?
- [x] Did you add your new functionality to the docs?
- [x] Did you update the [CHANGELOG](https://github.com/catalyst-team/catalyst/CHANGELOG.md)?
- [x] **You can use 'Login as guest' to see Teamcity build logs.**

## Description

Added full Metric Learning pipeline with example.

## Type of Change

- [ ] Examples / docs / tutorials / contributors update
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves an existing feature)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -52,13 +52,22 @@ pip install -U catalyst
 ```
 
 ```python
+#!/usr/bin/env python
 import os
+
+import numpy as np
+
+#!/usr/bin/env python
```
Comment (Scitator) [inline README.md]: bug

File: README.md [hunk 2]
```diff
@@ -177,9 +186,6 @@ Tested on Ubuntu 16.04/18.04/20.04, macOS 10.15, Windows 10 and Windows Subsyste
 <p>
 
 ```python
-import torch
```
Comment (Scitator) [inline README.md]: and here

File: README.md [hunk 3]
```diff
@@ -760,15 +711,6 @@ utils.distributed_cmd_run(train)
...
 ```python
-#!/usr/bin/env python
-import os
-import torch
-from torch import nn
-from torch.nn import functional as F
-from catalyst import dl, utils
-from catalyst.contrib.data.transforms import ToTensor
-from catalyst.contrib.datasets import MNIST
-from catalyst.utils import metrics
```
Comment (Scitator) [inline README.md]: up to here

File: README.md [hunk 4]
```diff
@@ -836,6 +778,70 @@ utils.distributed_cmd_run(train)
...
+from catalyst.contrib.datasets.mnist import MnistMLDataset, MnistQGDataset
+from catalyst.contrib.dl.callbacks import PeriodicLoaderCallback
+from catalyst.contrib.models.simple_conv import SimpleConv
+from catalyst.contrib.nn.criterion.triplet import TripletMarginLossWithSampling
+from catalyst.core.callbacks import ControlFlowCallback
+from catalyst.data.sampler import BalanceBatchSampler
+from catalyst.data.sampler_inbatch import AllTripletsSampler
+from catalyst.dl import CriterionCallback
+from catalyst.dl.callbacks.metrics.cmc import CMCScoreCallback
+from catalyst.dl.runner import SupervisedRunner
```
Comment (Scitator) [inline README.md]: let's just use `from catalyst import dl` and use Callbacks, Runners, like `dl.{something}`

Comment (AlekseySh) [reply README.md]: done

File: README.md [hunk 5]
```diff
@@ -836,6 +836,74 @@ utils.distributed_cmd_run(train)
...
+<summary>CV - MNIST with Metric Learning</summary>
+<p>
+
+```python
+from torch.optim import Adam
+from torch.utils.data import DataLoader
+
+from catalyst import contrib as ctb, data, dl
+from catalyst.core.callbacks import ControlFlowCallback
+from catalyst.dl.callbacks.metrics.cmc import CMCScoreCallback
```
Comment (Scitator) [inline README.md]: you should have access to them through `dl.{CallbackName}`

File: README.md [hunk 6]
```diff
@@ -836,6 +836,74 @@ utils.distributed_cmd_run(train)
...
+    model=model,
+    criterion=criterion,
+    optimizer=optimizer,
+    callbacks=callbacks,
+    loaders={"train": train_loader, "valid": val_loader},
+    minimize_metric=False,
+    verbose=True,
+    valid_loader="valid",
+    num_epochs=200,
+    main_metric="cmc_1",
```
Comment (Scitator) [inline README.md:889]: `cmc01`?

File: README.md [hunk 7]
```diff
@@ -836,6 +836,74 @@ utils.distributed_cmd_run(train)
...
+sampler_inbatch = data.sampler_inbatch.AllTripletsSampler(max_output_triplets=512)
+criterion = ctb.nn.criterion.triplet.TripletMarginLossWithSampling(
+    margin=0.5, sampler_inbatch=sampler_inbatch
+)
+
+# 4. training with catalyst Runner
+callbacks = [
+    ControlFlowCallback(dl.CriterionCallback(), loaders="train"),
+    ControlFlowCallback(CMCScoreCallback(topk_args=[1]), loaders="valid"),
+    dl.callbacks.PeriodicLoaderCallback(valid=600),
```
Comment (Scitator) [inline README.md]: you have only 200 epochs for train

File: catalyst/contrib/datasets/mnist.py [hunk 8]
```diff
@@ -1,5 +1,6 @@
 # flake8: noqa
```
Comment (Scitator) [inline catalyst/contrib/datasets/mnist.py:1]: could we remove noqa?

File: README.md (hunk 6 above)
Comment (AlekseySh) [reply README.md:889]: it is still 'cmc_1'

File: README.md (hunk 7 above)
Comment (AlekseySh) [reply README.md]: thx, readme is outdated, I will rework it after green tests

File: README.md (hunk 5 above)
Comment (AlekseySh) [reply README.md]: yes, but this block of code became ugly after using dl.bla-bla
```
# 4. training with catalyst Runner
callbacks = [
    ControlFlowCallback(dl.CriterionCallback(), loaders="train"),
    ControlFlowCallback(CMCScoreCallback(topk_args=[1]), loaders="valid"),
    dl.callbacks.PeriodicLoaderCallback(valid=600),
]
```
Is it okay?

Comment (AlekseySh) [reply README.md]: this one looks worse, imho 
```
    # 4. training with catalyst Runner
    callbacks = [
        ControlFlowCallback(dl.CriterionCallback(), loaders="train"),
        ControlFlowCallback(
            dl.callbacks.metrics.cmc.CMCScoreCallback(topk_args=[1]),
            loaders="valid",
        ),
        dl.callbacks.PeriodicLoaderCallback(valid=600),
    ]
```

File: catalyst/contrib/datasets/mnist.py (hunk 8 above)
Comment (AlekseySh) [reply catalyst/contrib/datasets/mnist.py:1]: ok

File: tests/_tests_scripts/dl_z_mvp_mnist_metric_learning.py [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+    # 3. criterion with triplets sampling
+    criterion = ctb.nn.criterion.triplet.TripletMarginLossWithSampling(
+        margin=0.5, sampler_inbatch=sampler_inbatch
+    )
+
+    # 4. training with catalyst Runner
+    callbacks = [
+        ControlFlowCallback(dl.CriterionCallback(), loaders="train"),
+        ControlFlowCallback(
+            dl.callbacks.metrics.cmc.CMCScoreCallback(topk_args=[1]),
```
Comment (Scitator) [inline tests/_tests_scripts/dl_z_mvp_mnist_metric_learning.py]: you can just use `dl.CMCScoreCallback`

File: tests/_tests_scripts/dl_z_mvp_mnist_metric_learning.py [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+
+from catalyst import contrib as ctb, data, dl
+import catalyst.contrib.data.transforms as t
+from catalyst.contrib.models.simple_conv import SimpleConv
+from catalyst.core.callbacks import ControlFlowCallback
+import catalyst.data.sampler_inbatch as si
+from catalyst.dl.callbacks.metrics.cmc import CMCScoreCallback
+
+
+def run_ml_pipeline(sampler_inbatch: si.InBatchTripletsSampler) -> float:
```
Comment (Scitator) [inline tests/_tests_scripts/dl_z_mvp_mnist_metric_learning.py]: btw, could we just use it as `data. InBatchTripletsSampler `? as far as you have imported `from catalyst import contrib as ctb, data, dl`

File: requirements/requirements.txt [hunk 11]
```diff
@@ -1,6 +1,6 @@
 # Framework must-haves
 numpy>=1.16.4
-torch>=1.1.0
+torch<=1.5
```
Comment (Scitator) [inline requirements/requirements.txt]: could you please return `torch>=1.1.0` we are already working on torch version issue

File: README.md (hunk 6 above)
Comment (Scitator) [reply README.md:889]: @elephantmipt 👀  ;)

File: README.md (hunk 5 above)
Comment (Scitator) [reply README.md]: it should be
```python
callbacks = [
        dl.ControlFlowCallback(dl.CriterionCallback(), loaders="train"),
        dl.ControlFlowCallback(dl.CMCScoreCallback(topk_args=[1]),loaders="valid"),
        dl.PeriodicLoaderCallback(valid=600),
    ]
```

Comment (AlekseySh) [reply README.md]: ah, ok, will try

Comment (AlekseySh) [reply README.md]: done

File: tests/_tests_scripts/dl_z_mvp_mnist_metric_learning.py (hunk 9 above)
Comment (AlekseySh) [reply tests/_tests_scripts/dl_z_mvp_mnist_metric_learning.py]: done

File: tests/_tests_scripts/dl_z_mvp_mnist_metric_learning.py (hunk 10 above)
Comment (AlekseySh) [reply tests/_tests_scripts/dl_z_mvp_mnist_metric_learning.py]: done

Comment (AlekseySh) [reply tests/_tests_scripts/dl_z_mvp_mnist_metric_learning.py]: done

File: catalyst/contrib/datasets/mnist.py (hunk 8 above)
Comment (AlekseySh) [reply catalyst/contrib/datasets/mnist.py:1]: no, it is still needed for copypasted mnist

File: catalyst/contrib/datasets/__init__.py [hunk 12]
```diff
@@ -1,11 +1,10 @@
 # flake8: noqa
-
 import logging
 
 from catalyst.tools import settings
 
-from catalyst.contrib.datasets.mnist import MNIST
-from catalyst.contrib.datasets.mnist_qg import MnistQGDataset
+from .metric_learning import MetricLearningTrainDataset, QueryGalleryDataset
```
Comment (Scitator) [inline catalyst/contrib/datasets/__init__.py:6]: use full path please

File: catalyst/contrib/nn/criterion/triplet.py [hunk 13]
```diff
@@ -1,10 +1,15 @@
 # flake8: noqa
```
Comment (Scitator) [inline catalyst/contrib/nn/criterion/triplet.py:1]: could you please remove `flake8: noqa`?

File: tests/_tests_scripts/dl_z_mvp_mnist_metric_learning.py [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+    # 4. training with catalyst Runner
+    callbacks = [
+        dl.ControlFlowCallback(dl.CriterionCallback(), loaders="train"),
+        dl.ControlFlowCallback(
+            dl.CMCScoreCallback(topk_args=[1]), loaders="valid"
+        ),
+        dl.PeriodicLoaderCallback(valid=10),
+    ]
+
+    runner = dl.SupervisedRunner(device="cuda:0" if is_available() else "cpu")
```
Comment (Scitator) [inline tests/_tests_scripts/dl_z_mvp_mnist_metric_learning.py]: use `utils.get_device()`

File: catalyst/contrib/datasets/__init__.py (hunk 12 above)
Comment (AlekseySh) [reply catalyst/contrib/datasets/__init__.py:6]: Are you sure? This is init file

File: tests/_tests_scripts/dl_z_mvp_mnist_metric_learning.py (hunk 14 above)
Comment (AlekseySh) [reply tests/_tests_scripts/dl_z_mvp_mnist_metric_learning.py]: done

File: catalyst/contrib/nn/criterion/triplet.py (hunk 13 above)
Comment (AlekseySh) [reply catalyst/contrib/nn/criterion/triplet.py:1]: this `noqa` is needed for another code in the module, not for mine

File: tests/_tests_scripts/dl_z_mvp_mnist_metric_learning.py [hunk 15]
```diff
@@ -0,0 +1,94 @@
+import os
+
+from torch.optim import Adam
+from torch.utils.data import DataLoader
+
+from catalyst import data, dl
+from catalyst.contrib import datasets, models, nn
+import catalyst.contrib.data.transforms as t
+from catalyst.utils import get_device
```
Comment (Scitator) [inline tests/_tests_scripts/dl_z_mvp_mnist_metric_learning.py]: so, looks like you can
```
from catalyst import data, dl, utils

utils.get_device()
```

it also looks like we need to simplify contrib for more flatten usage :)

Comment (AlekseySh) [reply tests/_tests_scripts/dl_z_mvp_mnist_metric_learning.py]: done
